File: merit-cvo/merit_cvo/nli.py
"""
NLI / factual-consistency leg — a 0..1 probability the source SUPPORTS the claim.

Two zero-cost paths, both permissive-licensed and offline-capable:
  * HTTP (default, no third-party deps): POST MERIT_NLI_URL {claim, source} -> {score}. Point it at the
    bundled `nli-server` (HHEM-2.1-Open) or any compatible scorer. Uses only the stdlib (urllib).
  * In-process (pip install "merit-cvo[hhem]"): set MERIT_NLI_MODEL to a HF model id and it loads HHEM
    locally. No server, no network.

Returns None if unconfigured or on any error — the NLI leg is additive evidence, never a hard failure.
"""
from __future__ import annotations

import json
import os
import urllib.request

_local_model = None  # cached in-process HHEM model


def nli_available(url: str | None = None) -> bool:
    return bool(url or os.getenv("MERIT_NLI_URL") or os.getenv("MERIT_NLI_MODEL"))


def _score_http(url: str, claim: str, source: str, timeout: float) -> float | None:
    body = json.dumps({"claim": claim, "source": source}).encode()
    req = urllib.request.Request(url, data=body, headers={"content-type": "application/json"}, method="POST")
    with urllib.request.urlopen(req, timeout=timeout) as r:
        if r.status != 200:
            return None
        data = json.loads(r.read().decode())
    s = data.get("score")
    return None if s is None else max(0.0, min(1.0, float(s)))


def _score_local(model_id: str, claim: str, source: str) -> float | None:
    global _local_model
    if _local_model is None:
        from transformers import AutoModelForSequenceClassification  # type: ignore
        _local_model = AutoModelForSequenceClassification.from_pretrained(model_id, trust_remote_code=True)
        _local_model.eval()
    # HHEM: predict([(premise=source, hypothesis=claim)]) -> P(support)
    return max(0.0, min(1.0, float(_local_model.predict([(source, claim)])[0])))


def score_nli(claim: str, source: str, *, url: str | None = None, timeout: float = 8.0) -> float | None:
    """0..1 support probability from the configured scorer, or None if unavailable / errored."""
    url = url or os.getenv("MERIT_NLI_URL")
    try:
        if url:
            return _score_http(url, claim, source, timeout)
        model_id = os.getenv("MERIT_NLI_MODEL")
        if model_id and model_id not in ("none", "custom-nli"):
            return _score_local(model_id, claim, source)
    except Exception:
        return None  # additive layer — never throw
    return None


def nli_model_tag(url: str | None = None) -> str:
    if os.getenv("MERIT_NLI_MODEL"):
        return os.environ["MERIT_NLI_MODEL"]
    return "custom-nli" if (url or os.getenv("MERIT_NLI_URL")) else "none"

File: lib/verify/nli.ts
/**
 * Pluggable NLI / factual-consistency scorer for the verification engine (M1).
 *
 * State-of-the-art citation verification decomposes a claim and checks entailment against the source with a
 * small, cheap, benchmarkable model — e.g. Vectara's HHEM-2.1-Open (T5) or MiniCheck-7B — which match or beat
 * far larger LLMs on claim-wise faithfulness at a fraction of the cost. We keep the model OUT of this repo and
 * behind an HTTP boundary so the engine stays dependency-light and offline-safe:
 *
 *   - Unconfigured (default): `scoreNLI` returns `null` → the engine falls back to the LLM judge (today's behavior).
 *   - Configured via `MERIT_NLI_URL`: POST {claim, source} → expect {score: 0..1} (probability the source
 *     SUPPORTS the claim). Point this at a local HHEM/MiniCheck server or a hosted factual-consistency API.
 *
 * This lets M1 ship + be benchmarked without bundling a model, and lets deployments plug in a real scorer later
 * with zero code change. Returning `null` on any failure is deliberate: the NLI layer is additive evidence, it
 * must never hard-fail a verification (the numeric verifier + LLM judge remain the backstops).
 */

export function nliAvailable(): boolean {
  return !!process.env.MERIT_NLI_URL;
}

/** 0..1 support probability from the configured scorer, or null if unavailable / errored. */
export async function scoreNLI(claim: string, source: string): Promise<number | null> {
  const url = process.env.MERIT_NLI_URL;
  if (!url) return null;
  try {
    const controller = new AbortController();
    const t = setTimeout(() => controller.abort(), Number(process.env.MERIT_NLI_TIMEOUT_MS || 8000));
    const r = await fetch(url, {
      method: "POST",
      headers: { "content-type": "application/json" },
      body: JSON.stringify({ claim, source }),
      signal: controller.signal,
    }).finally(() => clearTimeout(t));
    if (!r.ok) return null;
    const d = (await r.json()) as { score?: unknown };
    const s = Number(d?.score);
    return Number.isFinite(s) ? Math.max(0, Math.min(1, s)) : null;
  } catch {
    return null; // additive layer — never throw
  }
}

/** Model tag recorded on every verdict for reproducibility / benchmark versioning. */
export function nliModelTag(): string {
  return process.env.MERIT_NLI_MODEL || (nliAvailable() ? "custom-nli" : "none");
}
